src/server/pegasus_hotkey_collector.cpp

```diff
@@ -0,0 +1,279 @@
+        return false;
+    result = _fine_result;
+    return true;
+}
+
+void hotkey_collector::capture_msg_data(dsn::message_ex **requests, const int count)
+{
+    if (_collector_state.load() == STOP || _collector_state.load() == FINISH || count == 0) {
+        return;
+    }
+    for (int i = 0; i < count; i++) {
+        ::dsn::blob key;
+        if (requests[i] != nullptr) {
+            dsn::task_code rpc_code(requests[0]->rpc_code());
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_MULTI_PUT) {
+                dsn::apps::multi_put_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+                if (thrift_request.kvs.size() > 1) {
+                    for (int j = 0; j < thrift_request.kvs.size() - 2; j++) {
+                        capture_str_data(key.to_string());
+                    }
+                }
+                return;
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_INCR) {
+                dsn::apps::incr_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.key;
+                capture_blob_data(key);
+                return;
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_CHECK_AND_SET) {
+                dsn::apps::check_and_set_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+                return;
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_CHECK_AND_MUTATE) {
+                dsn::apps::check_and_mutate_request thrift_request;
+                unmarshall(requests[0], thrift_request);
+                requests[0]->restore_read();
+                key = thrift_request.hash_key;
+                capture_str_data(key.to_string());
+                return;
+            }
+            if (rpc_code == dsn::apps::RPC_RRDB_RRDB_PUT) {
+                dsn::apps::update_request thrift_request;
+                unmarshall(requests[i], thrift_request);
+                requests[i]->restore_read();
+                key = thrift_request.key;
+                capture_blob_data(key);
+                return;
+            }
+        }
+    }
+}
+
+void hotkey_collector::capture_multi_get_data(const ::dsn::apps::multi_get_request &request,
+                                              const ::dsn::apps::multi_get_response &resp)
+{
+    if (_collector_state.load() == STOP || _collector_state.load() == FINISH) {
+        return;
+    }
+    if (resp.kvs.size() != 0) {
+        for (const auto &iter : resp.kvs) {
+            capture_blob_data(request.hash_key);
+        }
```

Comment:
Add an int parameter for capture_blob_data to represent for the count of this key to avoid loop calls.

Reply:
In `capture_fine_data` we need to capture the full string data to ensure accuracy, it still has to loop call.

Reply:
It doesn't matter, you can overload `capture_blob_data `.